### PR TITLE
[ops, docs] feat: add 'auto' default for kernel selection

### PR DESCRIPTION
**Pull request by TimYangst**
### What does this PR do?

> Restores the pre-#678 "pick a fast kernel for me" ergonomics without bringing back the global `VEOMNI_USE_LIGER_KERNEL` env var. Every `*_implementation` field on `OpsImplementationConfig` (other than `attn_implementation`) now defaults to `"auto"`, which is resolved at config-parse time per device via a small `OpPolicy` table registered next to each kernel's existing `register_op` / `KERNEL_REGISTRY.register` call. Also renames the cross-entropy `"npu"` backend to `"chunk_loss"` and makes it device-agnostic so it can be the auto pick on both GPU and NPU.
>
> Builds on #678 (the typed `OpsImplementationConfig` refactor) and #689 (relax `moe_implementation` Literal to str). Minimises behavioural impact for users who upgrade — old yaml configs keep working through legacy aliases (`moe_implementation: fused → auto`, `cross_entropy_loss_implementation: npu → chunk_loss`) with deprecation warnings.

### Checklist Before Starting

- Search for relative PRs/issues and link here: builds on #678 (kernel selection refactor) and #689 (relax `moe_implementation` Literal to str).
- PR title follows `[{modules}] {type}: {description}` format (enforced by [check_pr_title.yml](.github/workflows/check_pr_title.yml))

### Test

> 30 new tests in `tests/ops/test_auto_resolution.py`; 116 total `tests/ops/` tests pass (existing `tests/ops/test_moe_hw_gate.py` patches updated to the new device-aware helper); `make quality` clean. Coverage:
>
> - Defaults: every non-attention field is `"auto"`; every field has an `OpPolicy`.
> - Auto resolution per device matches the matrix below for GPU / NPU / CPU.
> - Best-effort fallback: when `liger-kernel` (RMSNorm/RoPE/SwiGLU) or `triton` (load-balancing loss) is missing, auto picks degrade to `eager` with a `WARNING`.
> - Legacy aliases: `moe_implementation: fused` and `cross_entropy_loss_implementation: npu` resolve to their replacements with a deprecation warning.
> - Explicit overrides pass through unchanged for all fields; explicit `triton` on a host without it raises a clear `ValueError` at config-parse time (no silent fallback).
> - `install_loss_mapping("chunk_loss")` works; legacy `install_loss_mapping("npu")` still works (warns + routes to chunk_loss).
> - Resolved-summary log marks `(auto)` only on auto-resolved fields, never on explicit overrides.
> - **Regression guards** (added after reviewer feedback):
>   - The `load_balancing_loss` OpSlot KernelSpec is device-agnostic so NPU patchgen MoE models (qwen3_5_moe, qwen3_omni_moe) can bind it.
>   - `IS_NPU_AVAILABLE` reflects actual device presence, not just `find_spec("torch_npu")`.
>   - `register_op` rejects duplicate `config_field` registrations (the assumption was implicit; now it raises at import time).
>   - The fused-MoE legacy bind path uses the device-aware NPU check, so dual-stack hosts (CUDA + `torch_npu` package) keep the GPU `fused_triton` path.
>   - Non-MoE models (Llama, plain Qwen, ...) skip the legacy MoE patch entirely so the new auto default doesn't make them pay any MoE cost.

### API and Usage Example

> **YAML — minimal config now picks fast kernels automatically:**
>
> ```yaml
> model:
>   ops_implementation:
>     attn_implementation: flash_attention_2
>     # everything else defaults to "auto"
> ```
>
> On GPU this resolves to:
>
> ```
> Resolved ops implementation (device=gpu):
>   attn_implementation                = veomni_flash_attention_2_with_sp
>   cross_entropy_loss_implementation  = chunk_loss (auto)
>   load_balancing_loss_implementation = triton (auto)
>   moe_implementation                 = fused_triton (auto)
>   rms_norm_implementation            = liger_kernel (auto)
>   rotary_pos_emb_implementation      = liger_kernel (auto)
>   swiglu_mlp_implementation          = liger_kernel (auto)
> ```
>
> Mixing `auto` with explicit overrides is supported — the summary makes overrides obvious by omitting the `(auto)` marker:
>
> ```yaml
> model:
>   ops_implementation:
>     moe_implementation: eager   # explicit override, e.g. for debug
>     # everything else stays "auto"
> ```
>
> **Backward-compatible aliases (with deprecation warning):**
>
> | Old | New |
> |-----|-----|
> | `moe_implementation: fused` | `moe_implementation: auto` |
> | `cross_entropy_loss_implementation: npu` | `cross_entropy_loss_implementation: chunk_loss` |

### Auto Resolution Matrix

> Exhaustive table of how each field's `"auto"` value resolves. "Device" is determined by `_current_device_type()`:
>
> 1. `IS_NPU_AVAILABLE` (now: torch_npu installed AND `torch.npu.is_available()`) → `npu`
> 2. else `IS_CUDA_AVAILABLE` → `gpu`
> 3. else → `cpu`
>
> | Field | GPU + deps installed | GPU - deps | NPU + deps installed | NPU - deps | CPU host |
> |---|---|---|---|---|---|
> | `rms_norm_implementation` | `liger_kernel` | `eager` (warn) | `npu` | `eager` (warn) | `eager` |
> | `rotary_pos_emb_implementation` | `liger_kernel` | `eager` (warn) | `npu` | `eager` (warn) | `eager` |
> | `swiglu_mlp_implementation` | `liger_kernel` | `eager` (warn) | `eager` (no fused kernel) | `eager` | `eager` |
> | `cross_entropy_loss_implementation` | `chunk_loss` (no extra deps) | `chunk_loss` | `chunk_loss` | `chunk_loss` | `eager` |
> | `load_balancing_loss_implementation` | `triton` | `eager` (warn) | `triton` (via `triton-ascend`) | `eager` (warn) | `eager` |
> | `moe_implementation` | `fused_triton` | raises at apply time† | `fused_npu` | raises at apply time† | `eager` |
>
> † MoE auto trusts the device-keyed pick — its KernelSpec doesn't declare a `requires=` clause, so `apply_veomni_fused_moe_patch` raises a clear error if the kernel can't load. This matches the project's "no silent fallback for explicit selections" rule for the MoE path; if you want eager on a triton-less GPU, set `moe_implementation: eager` explicitly.
>
> **Legend** — "deps installed":
>
> | Backend | Required package |
> |---|---|
> | `liger_kernel` | `liger-kernel` |
> | `triton` (load-balancing) | `triton` (CUDA) or `triton-ascend` (NPU); both expose the `triton` import name |
> | `fused_triton` (MoE) | `triton` |
> | `fused_npu` / `npu` | `torch_npu` and an actual Ascend device |
>
> **Legacy aliases** (warn + auto-convert before the matrix above is consulted):
>
> | Field | Old value | New value |
> |---|---|---|
> | `moe_implementation` | `fused` | `auto` (then matrix above) |
> | `cross_entropy_loss_implementation` | `npu` | `chunk_loss` |

### Design & Code Changes

> See `docs/design/kernel_selection.md` ("Auto resolution") and `docs/design/unified_kernel_registry.md` (§3 + the OpPolicy subsection) for the full design. Specific changes:
>
> **New infrastructure**
>
> - `veomni/ops/config/auto_policy.py` — `OpPolicy` dataclass + `register_op_policy` / `list_op_policies`. Lives separately from `_OPS_REGISTRY` and `KERNEL_REGISTRY` so ops that exist in only one of them (cross-entropy, MoE) can still participate.
> - `HardwareRequirement.device_type` now accepts `None` for device-agnostic kernels (used by `chunk_loss` and the load-balancing loss `triton` KernelSpec).
> - New `is_triton_available()` / `is_npu_device_available()` helpers in `veomni.utils.{import_utils,device}` so the resolver can mock-test cleanly and so device-presence is no longer confused with package-presence.
>
> **Per-kernel policy registrations** (in each kernel's `__init__.py`):
>
> | Field | Auto policy | Legacy alias |
> |---|---|---|
> | `rms_norm_implementation` | gpu→`liger_kernel`, npu→`npu` | — |
> | `rotary_pos_emb_implementation` | gpu→`liger_kernel`, npu→`npu` | — |
> | `swiglu_mlp_implementation` | gpu→`liger_kernel` (npu→eager) | — |
> | `moe_implementation` | gpu→`fused_triton`, npu→`fused_npu` | `fused → auto` |
> | `cross_entropy_loss_implementation` | gpu→`chunk_loss`, npu→`chunk_loss` | `npu → chunk_loss` |
> | `load_balancing_loss_implementation` | gpu→`triton`, npu→`triton` | — |
>
> **Cross-entropy rename**
>
> - `KernelSpec(name="npu", hardware=HardwareRequirement(device_type="npu"))` → `KernelSpec(name="chunk_loss", hardware=HardwareRequirement(device_type=None))`. The kernel is pure `torch.func.grad_and_value` over the eager CE kernel and runs on either accelerator — only the registry name was overly specific.
> - `install_loss_mapping("npu")` is retained as a deprecated alias path (warns + routes to `chunk_loss`).
>
> **OpsImplementationConfig.__post_init__**
>
> - New `_resolve_auto()` pass: legacy alias rewrite (warning) → auto pick from `OpPolicy.auto_backends` → best-effort availability check via `_backend_requirements_met` (looks up `OpSpec.backends.requires` and verifies `liger_kernel` / `torch_npu` / `triton` are installed; warns + degrades to `eager` if not). Ops that aren't in OpSpec (CE, MoE) trust the device-keyed pick and let the kernel raise at apply time.
> - New `_log_resolved_summary()`: emits a single multi-line `Resolved ops implementation (device=...)` block at the end of `__post_init__`. Auto-resolved fields carry an `(auto)` marker so explicit overrides stand out. Width is computed only over fields that actually exist on the dataclass, so a future op policy added before its field is safe.
>
> **NPU / dual-stack correctness fixes** (follow-up commits responding to reviewer feedback):
>
> - `IS_NPU_AVAILABLE` now imports `torch_npu` (registers the `torch.npu` backend) and consults `torch.npu.is_available()` so it reflects actual device presence. Public `is_torch_npu_available()` keeps its package-only semantics (consistent with the function name); a new `is_npu_device_available()` exposes the device check. Without this, a CUDA dev host with `torch_npu` installed would mis-pick NPU backends.
> - `load_balancing_loss/triton` KernelSpec hardware switched from `device_type="gpu"` to `device_type=None`. NPU patchgen models (`qwen3_5_moe_npu_patch_gen_config.py`, `qwen3_omni_moe_*_patch_gen_config.py`) declare `OpSlot("load_balancing_loss", "standard")`; with the new auto default, `_bind_veomni_ops` would try to resolve `triton` against a GPU-only spec on NPU and crash at model-build time. The kernel itself is portable (`triton` and `triton-ascend` share the import name); availability is gated by the OpSpec `requires=("triton",)` clause + the auto resolver's eager fallback.
> - `apply_veomni_fused_moe_patch` + `is_fused_moe_available` + `is_quack_gemm_available` switched from the bare `torch_npu` package check to `is_npu_device_available`. Pre-fix, a CUDA host that merely had `torch_npu` installed would resolve `auto` → `fused_triton`, then crash inside the legacy MoE bind because the NPU rejection branch fired on the package alone. Now dual-stack hosts stay on the GPU path.
> - `build_foundation_model` skips `apply_moe_patch_transformers_v4` for non-MoE models. Pre-PR with `default="eager"` this was a no-op; with the new auto default the patch was firing for every Llama / plain Qwen config — wastefully importing triton and binding the global pointer that never gets called, and crashing outright on dual-stack hosts. New `_config_is_moe` helper inspects `num_experts` / `num_local_experts` / `moe_num_experts` on the config (and any `text_config` / `language_config` sub-config for VLM wrappers) and short-circuits the patch when no MoE expert dimension is present.
> - `register_op` now rejects duplicate `config_field` registrations. `_backend_requirements_met` walks `list_ops()` and stops at the first match for a given `config_field`; two ops sharing one would create non-deterministic resolution. The check makes the implicit assumption explicit and surfaces misregistrations at import time rather than at config-parse time.
>
> **Defaults**
>
> - 6 `*_implementation` fields default `"eager"` → `"auto"`.
> - `attn_implementation` keeps `"flash_attention_2"` (the explicit default already covers the common case; SP rewriting is unchanged).

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- [x] Applied pre-commit checks (`make quality` clean)
- [x] Added/updated documentation (`docs/design/kernel_selection.md`, `docs/design/unified_kernel_registry.md`, `docs/usage/arguments.md`)
- [x] Added tests to CI workflow (`tests/ops/test_auto_resolution.py`, 30 cases under the existing `tests/ops/` collection)

### CODE REVIEW

File: veomni/arguments/arguments_types.py [hunk 1]
```diff
@@ -724,10 +752,145 @@ def __post_init__(self):
...
+            if op.config_field != config_field:
+                continue
+            backend = op.backends.get(backend_name)
+            if backend is None:
+                return True
+            for pkg in backend.requires:
+                if pkg == "liger_kernel" and not is_liger_kernel_available():
+                    return False
+                if pkg == "torch_npu" and not is_torch_npu_available():
+                    return False
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/arguments/arguments_types.py:890]: Done

**Comment by TimYangst** [comment]
@gemini-code-assist  please take a review

File: veomni/arguments/arguments_types.py [hunk 2]
```diff
@@ -724,10 +752,146 @@ def __post_init__(self):
...
+                if pkg == "triton" and not is_triton_available():
+                    return False
+            return True
+        return True
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/arguments/arguments_types.py:894]: Done

**Comment by FoolPlayer** [comment]
Great! But why we need `auto` flag?
